project05/project05.rmd

```
@@ -0,0 +1,323 @@
+    
+    # Reverse the strings
+    aligned_seq1 = aligned_seq1[::-1]
+    aligned_seq2 = aligned_seq2[::-1]
+    
+    return aligned_seq1, aligned_seq2
+```
+
+```{python}
+def smith_waterman(seq1, seq2, match=1, mismatch=-1, gap=-1):
+    '''Smith-Waterman algorithm for local alignment
+    
+    Args:
+        seq1 (str): input seq 1
+        seq2 (str): input seq 2
+        match: default = +1
+        mismatch: default = -1
+        gap: default = -1
+    
+    Returns:
+        aligned_seq1 (str)
+        aligned_seq2 (str)
+        score_matrix (numpy array): scoring matrix
+    '''
+    # Step 1: Initialize matrix with zeros
+    # Matrix should be (len(seq1)+1) rows by (len(seq2)+1)columns
+    rows = len(seq1) + 1  # We add + 1 because we need an extra row at the top for the starting position
+    columns = len(seq2) + 1
+    # np.zeros() = numpy function that creates a grid filled with zeros
+    score_matrix = np.zeros((rows, columns), dtype=int)  # data type integer we want whole numbers not decimals
+    traceback_matrix = np.zeros((rows, columns), dtype=int)  # Matrix to hold direction
+    
+    # Step 2: Fill the matrices using cal_score 
+    for i in range(1, rows):  # Start at 1, skip row 0
+        for j in range(1, columns):  # Start at 1, skip column 0
+            score, direction = cal_score(score_matrix, seq1, seq2, i, j, match, mismatch, gap)
+            score_matrix[i][j] = score
+            traceback_matrix[i][j] = direction
+    
+    # Step 3: Find maximum score position
+    max_score = np.max(score_matrix)  # np.max() finds the maximum value in the matrix
+    max_pos = np.unravel_index(np.argmax(score_matrix), score_matrix.shape)  # np.unravel_index() finds the position (i, j) of that maximum
+    # np.argmax() finds the index of the max value (but as a single number)
+    # unravel_index converts that single number into (row, col) format
+    
+    aligned_seq1, aligned_seq2 = traceback(seq1, seq2, traceback_matrix, (6,7))
```

This hard-coded value might work for the single example in your comments (TACTTAG and CACATTAA), but it will fail for any other pair of sequences where the maximum score is at a different location.
You can try: 
aligned_seq1, aligned_seq2 = traceback(seq1, seq2, traceback_matrix, max_pos)

---

I believe that value was left in accidentally, probably to test the code before they added the max_score and max_pos variables. But yes, swap in the max_pos variable for those coordinates and this code is perfectly functional!